backends/tfhe-cuda-backend/cuda/include/device.h

```diff
@@ -48,15 +48,18 @@ uint32_t cuda_is_available();
 
 void *cuda_malloc(uint64_t size, uint32_t gpu_index);
 
-void *cuda_malloc_async(uint64_t size, cudaStream_t stream, uint32_t gpu_index);
+void *cuda_malloc_async(uint64_t size, cudaStream_t stream, uint32_t gpu_index,
+                        uint64_t *size_tracker, bool allocate_gpu_memory);
 
-void cuda_check_valid_malloc(uint64_t size, uint32_t gpu_index);
+bool cuda_check_valid_malloc(uint64_t size, uint32_t gpu_index);
 
 void cuda_memcpy_async_to_gpu(void *dest, const void *src, uint64_t size,
```

Comment:
The same suggestion about having two versions of each `cuda_memcpy_` method applies here. However, I'm not convinced we can't simply reorganize the scratch functions so that all operations that know they don't allocate memory remain within a conditional (checking the boolean). This would simplify a lot this PR and include the memcpy methods. Are you sure we need to modify so many functions to include that parameter?